util/unifyfs-stage/src/unifyfs-stage.c

```diff
@@ -0,0 +1,292 @@
+ * - stage in: moving files in pfs to unifyfs volume before user starts
+ *   application,
+ *   e.g., unifyfs start --stage-in=<manifest file>
+ * - stage out: moving files in the unifyfs volume to parallel file system
+ *   after user application completes,
+ *   e.g., unifyfs terminate --stage-out=<manifest file>
+ *
+ * Currently, we request users to pass the <manifest file> to specify target
+ * files to be transferred. The <manifest file> should list all target files
+ * and their destinations, line by line.
+ *
+ * This supports two transfer modes (although both are technically parallel):
+ *
+ * - serial: Each process will transfer a file. Data of a single file will
+ *   reside in a single compute node.
+ * - parallel (-p, --parallel): Each file will be split and transferred by all
+ *   processes. Data of a single file will be spread evenly across all
+ *   available compute nodes.
+ *
+ * TODO:
+ * Maybe later on, it would be better to have a size threshold. Based on the
+ * threshold, we can determine whether a file needs to transferred serially (if
+ * smaller than threshold), or parallelly.
+ */
+#include <config.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <getopt.h>
+#include <errno.h>
+#include <limits.h>
+#include <libgen.h>
+#include <mpi.h>
+
+#include "unifyfs_const.h"
+#include "unifyfs-stage.h"
+
+int rank;
+int total_ranks;
+int verbose;
+
+static int debug;
+static int checksum;
+static int mode;
+static char* manifest_file;
+static char* mountpoint = "/unifyfs";
+static char* share_dir;
+
+static unifyfs_stage_t _ctx;
+
+/**
+ * @brief create a status file for unifyfs cli to get niotified with the data
+ * stage result
+ *
+ * @param status 0 indicates success
+ *
+ * @return 0 on success, errno otherwise
+ */
+static int create_status_file(int status)
+{
+    char filename[PATH_MAX] = { 0, };
+    FILE* fp = NULL;
+    const char* msg = status ? "fail" : "success";
+
+    sprintf(filename, "%s/%s", share_dir, UNIFYFS_STAGE_STATUS_FILENAME);
+
+    fp = fopen(filename, "w");
+    if (!fp) {
+        fprintf(stderr, "failed to create %s (%s)\n",
+                        filename, strerror(errno));
+        return errno;
+    }
+
+    fprintf(fp, "%s\n", msg);
+
+    fclose(fp);
+
+    return 0;
+}
+
+static struct option long_opts[] = {
+    { "checksum", 0, 0, 'c' },
+    { "debug", 0, 0, 'd' },
+    { "help", 0, 0, 'h' },
+    { "mountpoint", 1, 0, 'm' },
+    { "parallel", 0, 0, 'p' },
+    { "share-dir", 1, 0, 's' },
+    { "verbose", 0, 0, 'v' },
+    { 0, 0, 0, 0 },
+};
+
+static char* short_opts = "cdhm:ps:v";
+
+static const char* usage_str =
+    "\n"
+    "Usage: %s [OPTION]... <manifest file>\n"
+    "\n"
+    "Transfer files between unifyfs volume and external file system.\n"
+    "The <manifest file> should contain list of files to be transferred,\n"
+    "and each line should be formatted as\n"
+    "\n"
+    "  /source/file/path,/destination/file/path\n"
```

Comment:
@MichaelBrim Nothing wrong with a space. I did this for avoiding test errors (mistakenly putting two spaces). We can switch to a space or any number of spaces and put extra code for parsing those tokens correctly.

Reply:
I like the spaces better too.  You will need to add addition logic to parse filenames with spaces though (either escaping the spaces `'\ '` or quoting the filename).

Reply:
Technically, POSIX portable pathnames (https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_282) don't allow spaces, and we could forbid them inside the UnifyFS namespace. But, we would still need to accommodate the external filenames that include spaces.

Reply:
My suggestion would be to require the use of quotes (either ' or ") surrounding filenames with spaces.

Reply:
We will follow the suggestions of using space. I used comma because it was simpler; I think there would are filenames with spaces and quotes than with comma (maybe I am wrong). From user's perspective, I am not sure if putting a space and required surrounding quotes would be more convenient than just putting a comma in between.

Reply:
The new version that we're testing now uses spaces, and in fact, the new one that Tony has sent me does this too.  So this is resolved.

Reply:
@craigsteffen the `usage_str` still needs to be updated to reflect the new manifest line format

Reply:
Oops.  This was so far up on the PR page I missed it today when I was cleaning up.  I'll do this now.

Reply:
This is now fixed in the force push I just pushed in.  It has a correct and slightly more elaborate usage message about the manifest file. 

I also put the MPI_Init() after the argument parsing, so that if it's invoked without an MPI environment with --help or no arguments, it produces sensible output.